[Draft] RFC: Patch dependencies using unidiff patchfiles

TotalKrill opened the pull request:
This being my first RFC for rust, I got some feedback that I should open a PR, even though the PR and the RFC is not completed. 

[rendered](https://github.com/TotalKrill/rfcs/blob/master/text/3177-apply-patchfiles-to-dependencies.md)

Feedback on the RFC itself, the writing and the process are all welcome. I find it easier to discuss around a PR ( thanks to githubs discussion system around them ) than trying to gather feedback on my fork. 

I realise that patchfiles are an incredible powerful tool, and the usecase I am presenting here might not be the optimal representation of this feature. But it was the solution I wanted when encountering this problem for the N-th time, where N > 3. 

This RFC draft is based upon discussions in: https://github.com/rust-lang/cargo/issues/4648 and https://github.com/rust-lang/cargo/issues/3830

TotalKrill commented:
I feel the Reference-level explanation is lacking, and would like it if someone helped out here

bjorn3 commented:
I think the patch file should be specified in the `[patch]` section too, not `[dependencies]`.

pickfire commented:
@TotalKrill can you please add a rendered link like other rfc pull requests?

TotalKrill commented:
@bjorn3 Generally I am a bit unhappy with the [patches] section, I feel it is hard to direct the patch to the specific dependency one would want changed.

roblabla commented:
This should definitely go in the `[patch]` section IMO. When working on workspaces, the `[patch]` section is centralized at the workspace level, ensuring all the members of the workspace use the same patched version of the package. I feel like this is super important. Furthermore, I feel that the proposed patchfile feature would actually fix some of the UX problems that the patch section has WRT it being nitpicky about the version it patches.

-----

FWIW, I have a slightly different motivation for such a feature. At work, our codebase has a bunch of crates in the `[patch]` section (around 20 currently) containing various bugfixes and whatnot that are in the progress of being upstreamed. We use github forks to maintain our patches, but the workflow around this is complicated, especially with regards to the long-term buildability of our project.

To ensure we can always rebuild old versions of our codebase, we need to ensure that the pinned hashes of our patches stay alive forever, which means making sure we never `push --force` those branches, and creating a new branch whenever we need to update the crate version.

The only current alternative to our current workflow is to vendor those patched dependencies, but this has its own set of downsides. Having patchfiles would be ideal for us: it'd trivially ensure old versions are reproducible, as we'd store the patchfiles in the same repo as the rest of the project.

TotalKrill commented:
@roblabla That is very valid points, and we are in the same boat basically.  If you go to my branch, I think you could expand the motivation text for this with those points, and also making the note about the workspace case, which I had forgotten, and Ill merge them and this RFC get that update?

TotalKrill commented:
I updated the RFC to move it into the `[patch]` section, that in turn added the requirement of a `version` key into said section, and I added some logic to handle that in a way that would not break how it is utilized today, as well as a suggestion for future work that could extend the usage of that new key

nagisa commented:
The RFC mentions unidiff as being a standardized format in some form or shape. You also mentioned “Looking at the unidiff standard” in one of your earlier comments. Is there a canonical description of this format anywhere? And if so, could it be linked from the RFC?

bjorn3 commented:
Can you mention that patching `Cargo.toml` is not supported? The `Cargo.toml` in a crate uploaded to crates.io is autogenerated and not the original one the user wrote. The original file stays as `Cargo.toml.orig`, but it may no longer be possible to create the autogenerated `Cargo.toml` from it even if it isn't patched.

TotalKrill commented:
>.... There's basically no way to patch a crates.io crate and if you're patching a git dependency you still have to know the exact _revision_, not just the branch/etc, to know what to patch ....

Long time since this was last discussed, had hoped some more interest would have gathered for this, but patching using patchfiles doesnt require any knowledge about git or anything in the usual case, it requires knowledge about where the file to patch is located, which lines, and some optionally configurable settings that can define that the previous lines should be something, or what the lines following the change should be.

Basically a patchfile is doing similar to what the changelog in merge requests does, it applies changes to files that looks a certain way. In no way does that require intricate knowledge of what git revision or the exact code it acts on. 

If the files the patchfile wants to change does not look like the patchfile specifies ( or the files does not exist ) then a clear error should be thrown during compilation. something along the lines of "patch XXXXX does not apply: {<filename> does not exist,  patchfile does not exist, lines 119-120 in <filename> does not match }"

da-x commented:
For prior art, here an equivalent functionality from NodeJS packaging context: [patch-package](https://yarnpkg.com/package/patch-package). It also uses unidiff files. Seems it is widely used (6k stars), and we can learn from its implementation and how to teach users the usage of this new functionality.

TotalKrill commented:
Here is a tool that has sprung up allowing patching. 

https://crates.io/crates/patch

I would like some kind of feedback on how to get this RFC moving towards a decision.

roblabla commented:
I've been thinking about this RFC, (which would really help us at work) and I think something we could do to push this forward is to experiment with this workflow out-of-tree, by (ab)using cargo vendor to build the dependencies. E.G. you could have a cargo plugin crate (say, `cargo build-patched`) that:

1. Calls `cargo vendor` to get all the deps in one place
2. Applies the patchfiles to the patched crates
3. Runs cargo build with a config override to build with the patched vendor directory.

It wouldn't be ideal (especially the whole "needs a custom build command" bit), but it'd allow us to work out the details of how the patches are generated and applied.

joshtriplett commented on text/3177-apply-patchfiles-to-dependencies.md:
```diff
@@ -0,0 +1,328 @@
+crates.io, path or git. 
+
+It would apply the patch from the patchfiles in the order they are defined in the patches section of
+the dependency declaration, if a patch does not apply cleanly, it would result in an compilation error
+that shows which patch that did not apply cleanly to which library. This would make sure that patchfiles are not to dangerous. 
```
```suggestion
the dependency declaration, if a patch has any "fuzz" (mismatched context lines), it would result in an compilation error
that shows which patch that did not apply cleanly to which library. This would make sure that patchfiles are not to dangerous. 
```

programmerjake replied:
@joshtriplett imho it should also specify other mismatches too otherwise it'd technically be legal to not fail if a patch deletes lines that don't match since they're not context lines

joshtriplett commented:
I'd love to see this happen. A few bits of feedback:

- This needs to specify the patch format.
- I think this should support `git-format-patch` patches, which provides some helpful extensions such as file moves/copies and patching binary files.
- Per my other comment, this should require that there be no fuzz.

cptpiepmatz commented:
What is the state on this?

loynoir commented:
Does this coverage use case like `cargo install XXX`?

ambasta commented:
How would this work if the patchfile patches the dependency Cargo.toml?

For example, adding feature flags to dependency that it doesn't expose. In this case, dependency resolution would fail to find a package with added features, which would need the patchfile to be applied before dependency resolution, which would in turn require dependency resolution.

linyihai commented:
If the current RFC is executed, can it reduce the size of the dependency library under certain circumstances? Sorry, I read the motivation section and I'm not sure what that means

ehuss commented:
I'm going to propose to postpone this RFC. The Cargo team agrees that this seems like it would be useful and a good idea. However, there are some details to work out, and this RFC has stalled somewhat. Primarily, there isn't anyone on the team who has the capacity at this time to champion this feature.

We would be interested in possibly seeing this land as an experimental feature, where we can explore options for the solution. This would be with the intent that an RFC would follow up before stabilization. If anyone is interested in working on that, feel free to reach out to the Cargo team on [#t-cargo](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo) on Zulip, or on the issue https://github.com/rust-lang/cargo/issues/4648, or come to [Office hours](https://github.com/rust-lang/cargo/wiki/Office-Hours).

@rfcbot fcp postpone
